activerecord/lib/active_record/associations/association.rb

```diff
@@ -321,6 +325,19 @@ def skip_statement_cache?(scope)
             klass.scope_attributes? ||
             reflection.source_reflection.active_record.default_scopes.any?
         end
+
+        def inversable?(record)
+          record &&
+            ((!record.persisted? || !owner.persisted?) || matches_foreign_key?(record))
+        end
+
+        def matches_foreign_key?(record)
+          if foreign_key_for?(record)
+            record.read_attribute(reflection.foreign_key) == owner.id
```

Comment:
This check isn't perfectly correct for polymorphic association.

Reply:
If I'm understanding correctly, this should also check to make sure the type matches in order to be completely correct? I have written some code locally to check for this type match in addition to the key match. But since we are always looking at reflections I struggle to setup a situation with a type mismatch to write a test against.

Reply:
I couldn't find a way to test for this being an issue. The issue this fixes comes from expanding the association scope using `.or` or similar, but I don't think we can expand it in a way to include a different table or primary key collisions (at least not that I know of in a way that's reasonably supported)